src/cedict-service/indexed-db-store.js

```diff
@@ -0,0 +1,228 @@
+/**
+ * @module IndexedDbStore
+ */
+import Store from '@lexisCs/cedict-service/store.js'
+
+/** A an IndexedDB store object */
+export default class IndexedDbStore extends Store {
+  constructor (configuration) {
+    super(configuration)
+    this._configuration = configuration
+    // DB object is not available when store object is created. It must be added later.
+    this._db = null
+  }
+
+  /**
+   * Checks if the configuration supplied has all the necessary information in it.
+   * If configuration is not valid it will throw an error indicating which check failed.
+   *
+   * @param {object} configuration - A JSON like configuration object.
+   */
+  static checkConfiguration (configuration) {
+    if (!configuration.name) throw new Error('A store name is missing from a configuration')
+    if (!configuration.primaryIndex) throw new Error('A primaryIndex tree is missing from a configuration')
+  }
+
+  /**
+   * Associates a store with an IndexedDB interface of a database where it is located.
+   *
+   * @param {IDBDatabase} db - An interface for connecting to IndexedDB.
+   * @returns {Store} A self-reference for chaining.
+   */
+  associateWith (db) {
+    this._db = db
+    return this
+  }
+
+  /**
+   * Creates a store. Can be run from `onupgradeneeded` callback only.
+   *
+   * @returns {Promise<undefined>|Promise<Error>} A promise that is resolved if store was created successfully
+   *          and is rejected if creation failed.
+   */
+  create () {
+    return new Promise((resolve, reject) => {
+      const options = this._configuration.primaryIndex.keyPath ? { keyPath: this._configuration.primaryIndex.keyPath } : undefined
+      const store = this._db.createObjectStore(this._configuration.name, options)
+      if (this._configuration.indexes) {
+        Object.values(this._configuration.indexes).forEach(idx => {
+          try {
+            store.createIndex(idx.name, idx.keyPath, { unique: idx.unique })
+          } catch (error) {
+            reject(error)
+          }
+        })
+      }
+      resolve()
+    })
+  }
+
+  /**
+   * Deletes all records from the store.
+   *
+   * @returns {Promise<undefined>|Promise<Error>} A promise that is resolved if all records were removed successfully
+   *          and is rejected if operation failed.
+   */
+  clear () {
+    return new Promise((resolve, reject) => {
+      let transaction = this._db.transaction(this._configuration.name, IndexedDbStore.accessModes.READ_WRITE) // eslint-disable-line prefer-const
+      let objectStore = transaction.objectStore(this._configuration.name) // eslint-disable-line prefer-const
+      let clearRequest = objectStore.clear() // eslint-disable-line prefer-const
+      clearRequest.onsuccess = () => resolve()
+      transaction.onerror = (event) => reject(event)
```

Comment:
does is matter that this isn't defined until after the transaction action is executed?

Reply:
Do you mean that the `objectStore` or `clearRequest` are not defined or that the `event` is not?
I did not find a very detailed description of how transactions works in IndexedDB (unless that it is "automatic") but from what I know:
- objectStore() request becomes a part of a transaction
- because clearRequest uses objectStore it becomes a part of a transaction too
- transaction completes automatically when all operations that are part of it are finished (i.e. getting an object store and clearing it)
- clearRequest.onsuccess will be called before any transaction handlers (onsuccess/onerror)

MDN example shows a similar approach: https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/clear. We just do not monitor transaction.onsuccess because I see no value in it.

Reply:
no, I mean that I think transaction.onerror should be set right after the transaction is defined, i.e. on line 69, rather than after the call to transaction.objectStore, otherwise I think it will have no effect, because it doesn't get defined until after the the code that might throw an error executes. That's the way the example code you referenced has it too.

Reply:
Ah, that makes sense, good catch! Will update.